contracts/src/v0.8/vrf/dev/OptimismL1Fees.sol

```diff
@@ -0,0 +1,97 @@
+
+/// @dev An abstract contract that provides Optimism specific L1 fee calculations.
+// solhint-disable-next-line contract-name-camelcase
+abstract contract OptimismL1Fees is ConfirmedOwner {
+  /// @dev This is the padding size for unsigned RLP-encoded transaction without the signature data
+  /// @dev The padding size was estimated based on hypothetical max RLP-encoded transaction size
+  uint256 internal constant L1_UNSIGNED_RLP_ENC_TX_DATA_BYTES_SIZE = 71;
+  /// @dev Signature data size used in the GasPriceOracle predeploy
+  /// @dev reference: https://github.com/ethereum-optimism/optimism/blob/a96cbe7c8da144d79d4cec1303d8ae60a64e681e/packages/contracts-bedrock/contracts/L2/GasPriceOracle.sol#L145
+  uint256 internal constant L1_TX_SIGNATURE_DATA_BYTES_SIZE = 68;
+  /// @dev L1_FEE_DATA_PADDING includes 71 bytes for L1 data padding for Optimism
+  bytes internal constant L1_FEE_DATA_PADDING =
+    hex"ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff";
+  /// @dev OVM_GASPRICEORACLE_ADDR is the address of the OVM_GasPriceOracle precompile on Optimism.
+  /// @dev reference: https://community.optimism.io/docs/developers/build/transaction-fees/#estimating-the-l1-data-fee
+  address private constant OVM_GASPRICEORACLE_ADDR = address(0x420000000000000000000000000000000000000F);
+  OVM_GasPriceOracle private constant OVM_GASPRICEORACLE = OVM_GasPriceOracle(OVM_GASPRICEORACLE_ADDR);
+
+  /// @dev Option 1: getL1Fee() function from predeploy GasPriceOracle contract with the fulfillment calldata payload
+  /// @dev This option is only available for the Coordinator contract
+  uint8 internal constant L1_GAS_FEES_MODE = 0;
+  /// @dev Option 2: our own implementation of getL1Fee() function (Ecotone version) with projected
+  /// @dev fulfillment calldata payload (number of non-zero bytes estimated based on historical data)
+  /// @dev This option is available for the Coordinator and the Wrapper contract
+  uint8 internal constant L1_CALLDATA_GAS_COST_MODE = 1;
+  /// @dev Option 3: getL1FeeUpperBound() function from predeploy GasPriceOracle contract (available after Fjord upgrade)
+  /// @dev This option is available for the Coordinator and the Wrapper contract
+  uint8 internal constant L1_GAS_FEES_UPPER_BOUND_MODE = 2;
+
+  uint8 public s_l1FeeCalculationMode = L1_GAS_FEES_MODE;
+
+  /// @dev L1 fee coefficient can be applied to options 2 or 3 to reduce possibly inflated gas price
+  uint8 public s_l1FeeCoefficient = 100;
+
+  error InvalidL1FeeCalculationMode(uint8 mode);
+  error InvalidL1FeeCoefficient(uint8 coefficient);
+
+  event L1FeeCalculationSet(uint8 mode, uint8 coefficient);
+
+  function setL1FeeCalculation(uint8 mode, uint8 coefficient) external virtual onlyOwner {
+    _setL1FeeCalculationInternal(mode, coefficient);
```

Comment:
I wanted to override `setL1FeeCalculation` in the `VRFV2PlusWrapper_Optimism` contract to make sure that `L1_GAS_FEES_MODE` can't be used. But at the same time, I wanted to use the same code from the `OptimismL1Fees`  base contract. There was no other way to achieve this but to create an internal function to wrap up this code inside.

Reply:
Not sure if I'm following, but why can't we just inline the `_setL1FeeCalculationInternal` into `setL1FeeCalculation`? There's no difference right? Overriding `setL1FeeCalculation` in `VRFV2PlusWrapper_Optimism` will still replace the base function.

Reply:
That's what I also expected, but it doesn't work if you're trying to override the external function this way.

So this is how it looked like before:
```
function setL1FeeCalculation(uint8 mode, uint8 coefficient) external virtual onlyOwner {
    if (mode >= 3) {
      revert InvalidL1FeeCalculationMode(mode);
    }
    if (coefficient == 0 || coefficient > 100) {
      revert InvalidL1FeeCoefficient(coefficient);
    }

    s_l1FeeCalculationMode = mode;
    s_l1FeeCoefficient = coefficient;

    emit L1FeeCalculationSet(mode, coefficient);
  }
```

And when I tried to override this in the wrapper contract for Optimism:
```
function setL1FeeCalculation(uint8 mode, uint8 coefficient) external override onlyOwner {
    if (mode == L1_CALLDATA_GAS_COST_MODE) {
      revert InvalidL1FeeCalculationMode(mode);
    }
    OptimismL1Fees.setL1FeeCalculation(mode, coefficient);
  }
```

But it didn't work because the compiler was complaining about it. Something like this was the error:
Member "setL1FeeCalculation" not found or not visible after argument-dependent lookup